src/ledger/LedgerTxnOfferSQL.cpp

```diff
@@ -895,4 +675,131 @@ LedgerTxnRoot::Impl::bulkLoadOffers(
+    auto prep = db.getPreparedStatement(sql);
+
+    std::string actIDStrKey;
+    unsigned int sellingAssetType, buyingAssetType;
+    std::string sellingAssetCode, buyingAssetCode, sellingIssuerStrKey,
+        buyingIssuerStrKey;
+    soci::indicator sellingAssetCodeIndicator, buyingAssetCodeIndicator,
+        sellingIssuerIndicator, buyingIssuerIndicator;
+
+    LedgerEntry le;
+    le.data.type(OFFER);
+    OfferEntry& oe = le.data.offer();
+
+    auto& st = prep.statement();
+    st.exchange(soci::into(actIDStrKey));
+    st.exchange(soci::into(oe.offerID));
+    st.exchange(soci::into(sellingAssetType));
+    st.exchange(soci::into(sellingAssetCode, sellingAssetCodeIndicator));
+    st.exchange(soci::into(sellingIssuerStrKey, sellingIssuerIndicator));
+    st.exchange(soci::into(buyingAssetType));
+    st.exchange(soci::into(buyingAssetCode, buyingAssetCodeIndicator));
+    st.exchange(soci::into(buyingIssuerStrKey, buyingIssuerIndicator));
+    st.exchange(soci::into(oe.amount));
+    st.exchange(soci::into(oe.price.n));
+    st.exchange(soci::into(oe.price.d));
+    st.exchange(soci::into(oe.flags));
+    st.exchange(soci::into(le.lastModifiedLedgerSeq));
+    st.define_and_bind();
+    st.execute(true);
+    while (st.got_data())
+    {
+        oe.sellerID = KeyUtils::fromStrKey<PublicKey>(actIDStrKey);
+        processAsset(oe.selling, (AssetType)sellingAssetType,
+                     sellingIssuerStrKey, sellingIssuerIndicator,
+                     sellingAssetCode, sellingAssetCodeIndicator);
+        processAsset(oe.buying, (AssetType)buyingAssetType, buyingIssuerStrKey,
+                     buyingIssuerIndicator, buyingAssetCode,
+                     buyingAssetCodeIndicator);
+
+        offers.emplace_back(le);
+        st.fetch();
+    }
+
+    return offers;
+}
+
+void
+LedgerTxnRoot::Impl::writeOffersIntoSimplifiedOffersTable()
+{
+    throwIfChild();
+    mEntryCache.clear();
+    mBestOffersCache.clear();
+
+    CLOG(INFO, "Ledger") << "Loading all offers";
+    auto const offers = stellar::loadAllOffers(mDatabase);
+
+    mDatabase.getSession() << "DROP TABLE IF EXISTS offers";
+    mDatabase.getSession()
+        << "CREATE TABLE offers"
+           "("
+           "sellerid         VARCHAR(56)      NOT NULL,"
+           "offerid          BIGINT           NOT NULL CHECK (offerid >= 0),"
+           "sellingasset     TEXT             NOT NULL,"
+           "buyingasset      TEXT             NOT NULL,"
+           "amount           BIGINT           NOT NULL CHECK (amount >= 0),"
+           "pricen           INT              NOT NULL,"
+           "priced           INT              NOT NULL,"
+           "price            DOUBLE PRECISION NOT NULL,"
+           "flags            INT              NOT NULL,"
+           "lastmodified     INT              NOT NULL,"
+           "PRIMARY KEY      (offerid)"
+           ");";
+    mDatabase.getSession() << "CREATE INDEX bestofferindex ON offers "
```

Comment:
This is a different set of indexes than the previous table and I'm slightly concerned that it will either fail to support the current or future set of queries; in particular AllowTrustOp seems like it does point queries that will want an index on each of buyingasset and sellingasset (it had them before on the issuers-component of each); and I was unclear on whether horizon wanted to do point queries on the table by asset or account+asset too.

(Also: for indexes that are supporting point queries, we should probably use hash indexes rather than the default btrees)

Reply:
Why include `offerID` in that index? `offerID` is just a tie breaker, and the assumption is that prices tend to be different. I would add this one later if we need to.

What about the index on `buyingasset`? without it, `loadOffersByAccountAndAsset` will perform a full scan on the entire order book.

As for future uses, I am not too worried about that: I'd rather have a schema that reflects are current patterns.

For Horizon, this preserves indexes for path finding (that looks at selling or buying assets). Other queries are already bad in Horizon as they query on `sellerID` afaik.

Reply:
I'm not sure that assumption is right. A quick query of the database (as of ledger approximately 20 million) suggests there are

- 1312 offers where `offerID` would be needed to differentiate between 2
- 459 offers where `offerID` would be needed to differentiate between 3
- 220 offers where `offerID` would be needed to differentiate between 4
- 1667 offers where `offerID` would be needed to differentiate between 5 or more

and a specific example is 880 offers with the same `(sellingasset, buyingasset, price)`. All together, 31.7% of offers have the same `(sellingasset, buyingasset, price)` as at least one other offer.

With regard to `loadOffersByAccountAndAsset`, I'm not even sure if an index on `buyingasset` would be sufficient. We'd probably want to have an index on `(sellerid, sellingasset)` and an index on `(sellerid, buyingasset)` if we actually wanted the query to be well-optimized. I will look tomorrow into how that query performs with different kinds of index and reply with the results.

Reply:
any update on this?

I would drop `offerID` for now
* we don't have this index right now
* 99% of the database has less than 5 offers at any given price, so it doesn't seem that speeding up the 1% is worth it at this time (while slowing down the 99%), unless you have data that shows that it makes any difference (like, replay ranges of legers instead of synthetic tests)

re: `loadOffersByAccountAndAsset` is probably a similar situation where 99% of the users have very few offers per trading pair so a small scan is probably fine

Reply:
I have run some artificial benchmarks, and also a small benchmark that catches up at ledger 22.5 million then replays 5000 ledgers on sqlite.

With regard to whether the index should contain `offerID`, it was faster in the benchmarks. But the actual history showed it was about 25% slower for bucket apply and effectively unchanged for replay. I think we should exclude `offerID` until the data in the ledger necessitates inclusion.

With regard to `loadOffersByAccountAndAsset` I tried adding various potentially useful indexes. In artificial benchmarks they do make this query a bit faster but at the cost of making writes much slower. In actual history, all indexes I tried made both bucket apply and replay slower. It's possible, perhaps even likely, that this range of ledgers did not include that query even once. But given the observations, I also think we should not add additional indexes for this purpose until we have concrete data supporting the effectiveness.

In summary, my proposal is remove `offerid` from `bestofferindex` and add no other indexes. @graydon @MonsieurNicolas if you are okay with this, I will update the PR accordingly.